// @riadh-adrani/utility-js/build/src/color/generate/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.changeColorOpacity = exports.generateColorTonalPalette = exports.generateComplementaryColor = exports.generateContrastSafeColor = void 0;
const getType_1 = require("../getType");
const convert_1 = require("../convert");
const rgb_1 = require("../rgb");
const hsl_1 = require("../hsl");
const math_1 = require("../../math");
function normalizeColorToRgbOrThrow(color) {
    const rgb = (0, convert_1.convertColor)(color, "rgb");
    const t = (0, getType_1.getColorType)(rgb);
    if (t === "unknown") {
        throw "unable to specify (color) type, maybe it is badly formatted ?";
    }
    return rgb;
}
function normalizeColorToHSLOrThrow(color) {
    const rgb = (0, convert_1.convertColor)(color, "hsl");
    const type = (0, getType_1.getColorType)(rgb);
    if (type === "unknown") {
        throw "unable to specify (color) type, maybe it is badly formatted ?";
    }
    return rgb;
}
/**
 * generates a safe color that could be paired with the given one,
 * as a text or background color.
 *
 * @param color source
 * @param light color for dark colors
 * @param dark color for light colors
 */
function generateContrastSafeColor(color, light = "#fff", dark = "#000") {
    const c = normalizeColorToRgbOrThrow(color);
    const [r, g, b] = (0, rgb_1.extractDataFromRGB)(c);
    const yiq = (r * 299 + g * 587 + b * 114) / 1000;
    return yiq >= 128 ? dark : light;
}
exports.generateContrastSafeColor = generateContrastSafeColor;
/**
 * generates a complementary color.
 * @param color source
 * @param type result color type
 */
function generateComplementaryColor(color, type) {
    const c = normalizeColorToHSLOrThrow(color);
    const [_h, _s, _l, _a] = (0, hsl_1.extractDataFromHSL)(c);
    const [h, s, l, a] = [(_h + 180) % 360, _s, _l, _a !== null && _a !== void 0 ? _a : 1];
    const result = `hsla(${h}deg ${s}% ${l}% / ${a})`;
    return (0, convert_1.convertColor)(result, type || "hsl");
}
exports.generateComplementaryColor = generateComplementaryColor;
/**
 * generates a tonal palette.
 * @param color
 * @param type output colors type
 * @see https://m3.material.io/styles/color/the-color-system/key-colors-tones#6bdb9471-b70d-42c9-8ace-76743c1fff13
 */
function generateColorTonalPalette(color, type) {
    const c = normalizeColorToHSLOrThrow(color);
    const palette = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 99, 100];
    const [h, s, , a] = (0, hsl_1.extractDataFromHSL)(c);
    const t = type !== null && type !== void 0 ? type : "hsl";
    const colorPalette = {};
    palette.forEach((tone) => {
        const colorTone = (0, hsl_1.hsla)(h, s, tone, a !== null && a !== void 0 ? a : 1);
        colorPalette[tone] = (0, convert_1.convertColor)(colorTone, t);
    });
    return colorPalette;
}
exports.generateColorTonalPalette = generateColorTonalPalette;
/**
 * generate the same color with a modified opacity.
 * @param color source
 * @param opacity opacity
 * @param type output color type
 */
function changeColorOpacity(color, opacity, type) {
    const c = normalizeColorToHSLOrThrow(color);
    if (!(0, math_1.isInInterval)(0, opacity, 1))
        throw "(opacity) should be a number between 0 and 1.";
    const [h, s, l] = (0, hsl_1.extractDataFromHSL)(c);
    return (0, convert_1.convertColor)((0, hsl_1.hsla)(h, s, l, opacity), type !== null && type !== void 0 ? type : "hex");
}
exports.changeColorOpacity = changeColorOpacity;
