// @riadh-adrani/utility-js/build/src/color/rgb/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRgbColor = exports.extractDataFromRGB = exports.isRgbaForm = exports.isRgbForm = exports.RGBA_REGEX = exports.RGB_REGEX = exports.rgba = void 0;
const string_1 = require("../../string");
const math_1 = require("../../math");
/**
 * generate an rgba color :
 *
 * `rgba(255,255,255,1)`
 *
 * @param r red
 * @param g green
 * @param b blue
 * @param a alpha
 */
function rgba(r, g, b, a = 1) {
    if (!(0, math_1.isInInterval)(0, r, 255))
        throw "(red) should be a number between 0 and 255";
    if (!(0, math_1.isInInterval)(0, b, 255))
        throw "(green) should be a number between 0 and 255";
    if (!(0, math_1.isInInterval)(0, g, 255))
        throw "(blue) should be a number between 0 and 255";
    if (!(0, math_1.isInInterval)(0, a, 1))
        throw "(alpha) should be a number between 0 and 1";
    return `rgba(${r},${g},${b},${a})`;
}
exports.rgba = rgba;
exports.RGB_REGEX = /^rgb\((\d{1,3})(\.\d+){0,1},( ){0,1}(\d{1,3})(\.\d+){0,1},( ){0,1}(\d{1,3})(\.\d+){0,1}\)$/;
exports.RGBA_REGEX = /^rgba\((\d{1,3})(\.\d+){0,1},( ){0,1}(\d{1,3})(\.\d+){0,1},( ){0,1}(\d{1,3})(\.\d+){0,1}(,( ){0,1}(\d{1})(\.\d+){0,1}){0,1}\)$/;
/**
 * check if the given color is of an rgb form
 *
 * `rgb(255, 255, 255)`
 *
 * note that red, green and blue values are not checked
 * @param color target
 */
function isRgbForm(color) {
    return (0, string_1.isBlank)(color) ? false : exports.RGB_REGEX.test(color);
}
exports.isRgbForm = isRgbForm;
/**
 * check if the given color is of an rgba form
 *
 * `rgba(255, 255, 255, 0.99)`
 *
 * note that red, green, blue and alpha values are not checked.
 * @param color target
 */
function isRgbaForm(color) {
    return (0, string_1.isBlank)(color) ? false : exports.RGBA_REGEX.test(color);
}
exports.isRgbaForm = isRgbaForm;
/**
 * extract color data from an rgb/a color form.
 *
 * if the color has an invalid type, an error will be thrown.
 *
 * @param color source
 */
function extractDataFromRGB(color) {
    if (isRgbaForm(color)) {
        return color
            .slice(5, -1)
            .split(",")
            .map((i) => parseFloat(i.trim()));
    }
    if (isRgbForm(color)) {
        return color
            .slice(4, -1)
            .split(",")
            .map((i) => parseFloat(i.trim()));
    }
    throw "(color) is not of a RGB/RGBA form.";
}
exports.extractDataFromRGB = extractDataFromRGB;
/**
 * Check if the given color is in an RGB or RGBA format:
 * - `rgb(r,g,b)`
 * - `rgba(r,g,b,a)`
 * @param color
 */
function isRgbColor(color) {
    if ((0, string_1.isBlank)(color)) {
        return false;
    }
    if (isRgbForm(color)) {
        const [r, g, b] = extractDataFromRGB(color);
        for (const c of [r, g, b]) {
            if (!(0, math_1.isInInterval)(0, c, 256))
                return false;
        }
        return true;
    }
    else if (isRgbaForm(color)) {
        const [r, g, b, a] = extractDataFromRGB(color);
        for (const c of [r, g, b]) {
            if (!(0, math_1.isInInterval)(0, c, 256))
                return false;
        }
        if (a !== undefined && !(0, math_1.isInInterval)(0, a, 1))
            return false;
        return true;
    }
    return false;
}
exports.isRgbColor = isRgbColor;
