// @riadh-adrani/utility-js/build/src/time/since/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.since = exports.yearsSince = exports.monthsSince = exports.weeksSince = exports.daysSince = exports.hoursSince = exports.minutesSince = exports.secondsSince = exports.timeUnitSince = exports.YEAR = exports.MONTH = exports.WEEK = exports.DAY = exports.HOUR = exports.MINUTE = exports.SECOND = void 0;
exports.SECOND = 1000;
exports.MINUTE = exports.SECOND * 60;
exports.HOUR = exports.MINUTE * 60;
exports.DAY = exports.HOUR * 24;
exports.WEEK = exports.DAY * 7;
exports.MONTH = exports.DAY * 30;
exports.YEAR = exports.DAY * 365;
/**
 * Count the time since date with a given time of unit (in milliseconds).
 * @param time date as a number.
 * @param timeUnit the unit of time in milliseconds.
 * @param singular the word to be appended when the result is singular.
 * @param plural the word to be appended when the result is plural.
 */
function timeUnitSince(time, timeUnit, singular, plural) {
    if (typeof time !== "number") {
        throw new Error("time is not a number.");
    }
    if (time > Date.now()) {
        throw new Error("the given time is superior to the current time.");
    }
    if (typeof timeUnit !== "number" || timeUnit < 0) {
        throw new Error("time unit is not a positive number.");
    }
    const delta = Date.now() - time;
    const count = Math.trunc(delta / timeUnit);
    return {
        count,
        formatted: `${count} ${count < 2 ? singular : plural}`,
    };
}
exports.timeUnitSince = timeUnitSince;
/**
 * Count the time since a given date in seconds.
 * @param time date.
 * @param plural `seconds` by default
 * @param singular `second` by default
 */
function secondsSince(time, singular = "second", plural = "seconds") {
    return timeUnitSince(time, exports.SECOND, singular, plural);
}
exports.secondsSince = secondsSince;
/**
 * Count the time since a given date in minutes.
 * @param time date.
 * @param plural `minutes` by default
 * @param singular `minute` by default
 */
function minutesSince(time, singular = "minute", plural = "minutes") {
    return timeUnitSince(time, exports.MINUTE, singular, plural);
}
exports.minutesSince = minutesSince;
/**
 * Count the time since a given date in hours.
 * @param time date.
 * @param plural `hours` by default
 * @param singular `hour` by default
 */
function hoursSince(time, singular = "hour", plural = "hours") {
    return timeUnitSince(time, exports.HOUR, singular, plural);
}
exports.hoursSince = hoursSince;
/**
 * Count the time since a given date in days.
 * @param time date.
 * @param plural `days` by default
 * @param singular `day` by default
 */
function daysSince(time, singular = "day", plural = "days") {
    return timeUnitSince(time, exports.DAY, singular, plural);
}
exports.daysSince = daysSince;
/**
 * Count the time since a given date in weeks.
 * @param time date.
 * @param plural `weeks` by default
 * @param singular `week` by default
 */
function weeksSince(time, singular = "week", plural = "weeks") {
    return timeUnitSince(time, exports.WEEK, singular, plural);
}
exports.weeksSince = weeksSince;
/**
 * Count the time since a given date in months.
 * @param time date.
 * @param plural `months` by default
 * @param singular `month` by default
 */
function monthsSince(time, singular = "month", plural = "months") {
    return timeUnitSince(time, exports.MONTH, singular, plural);
}
exports.monthsSince = monthsSince;
/**
 * Count the time since a given date in years.
 * @param time date.
 * @param plural `years` by default
 * @param singular `year` by default
 */
function yearsSince(time, singular = "year", plural = "years") {
    return timeUnitSince(time, exports.YEAR, singular, plural);
}
exports.yearsSince = yearsSince;
/**
 * Count the time since a given date an return the appropriate format with additional information.
 * @param time date.
 * @param _second array of [`singular`,`plural`]
 * @param _minute array of [`singular`,`plural`]
 * @param _hour array of [`singular`,`plural`]
 * @param _day array of [`singular`,`plural`]
 * @param _week array of [`singular`,`plural`]
 * @param _month array of [`singular`,`plural`]
 * @param _year array of [`singular`,`plural`]
 */
function since(time, _second = ["second", "seconds"], _minute = ["minute", "minutes"], _hour = ["hour", "hours"], _day = ["day", "days"], _week = ["week", "weeks"], _month = ["month", "months"], _year = ["year", "years"]) {
    if (typeof time !== "number") {
        throw new Error("time is not a number.");
    }
    let t = time;
    const years = yearsSince(t, ..._year);
    t += years.count * exports.YEAR;
    const months = monthsSince(t, ..._month);
    t += months.count * exports.MONTH;
    const weeks = weeksSince(t, ..._week);
    t += weeks.count * exports.WEEK;
    const days = daysSince(t, ..._day);
    t += days.count * exports.DAY;
    const hours = hoursSince(t, ..._hour);
    t += hours.count * exports.HOUR;
    const minutes = minutesSince(t, ..._minute);
    t += minutes.count * exports.MINUTE;
    const seconds = secondsSince(t, ..._second);
    const res = {
        formatted: "",
        seconds,
        minutes,
        hours,
        days,
        weeks,
        months,
        years,
    };
    res.formatted = (() => {
        if (years.count > 0)
            return years.formatted;
        else if (months.count > 0)
            return months.formatted;
        else if (weeks.count > 0)
            return weeks.formatted;
        else if (days.count > 0)
            return days.formatted;
        else if (hours.count > 0)
            return hours.formatted;
        else if (minutes.count > 0)
            return minutes.formatted;
        else
            return seconds.formatted;
    })();
    return res;
}
exports.since = since;
