// @riadh-adrani/utility-js/build/src/object/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.copy = exports.verify = exports.merge = exports.isPrimitiveType = exports.isFunction = exports.isFalsy = exports.hasProperty = exports.areEqual = void 0;
var areEqual_1 = require("./areEqual");
Object.defineProperty(exports, "areEqual", { enumerable: true, get: function () { return __importDefault(areEqual_1).default; } });
var hasProperty_1 = require("./hasProperty");
Object.defineProperty(exports, "hasProperty", { enumerable: true, get: function () { return __importDefault(hasProperty_1).default; } });
var isFalsy_1 = require("./isFalsy");
Object.defineProperty(exports, "isFalsy", { enumerable: true, get: function () { return __importDefault(isFalsy_1).default; } });
var isFunction_1 = require("./isFunction");
Object.defineProperty(exports, "isFunction", { enumerable: true, get: function () { return __importDefault(isFunction_1).default; } });
var isPrimitiveType_1 = require("./isPrimitiveType");
Object.defineProperty(exports, "isPrimitiveType", { enumerable: true, get: function () { return __importDefault(isPrimitiveType_1).default; } });
var merge_1 = require("./merge");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return __importDefault(merge_1).default; } });
var verify_1 = require("./verify");
Object.defineProperty(exports, "verify", { enumerable: true, get: function () { return __importDefault(verify_1).default; } });
var copy_1 = require("./copy");
Object.defineProperty(exports, "copy", { enumerable: true, get: function () { return __importDefault(copy_1).default; } });
__exportStar(require("./isTrue"), exports);
