// @riadh-adrani/utility-js/build/src/object/isTrue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDefined = exports.isNull = exports.isObject = exports.isArray = exports.isString = exports.isNumber = exports.isTrue = void 0;
/**
 * checks if the condition provided is true.
 * @param condition comparison or function.
 */
function isTrue(condition) {
    if (typeof condition === "boolean") {
        return condition;
    }
    if (typeof condition !== "function") {
        throw "(condition) should be a boolean or a function";
    }
    const res = condition();
    if (typeof res !== "boolean") {
        throw "(condition) result is not a boolean.";
    }
    return res;
}
exports.isTrue = isTrue;
/**
 * checks if the object provided is a number.
 * @param o
 */
function isNumber(o) {
    return isTrue(typeof o === "number");
}
exports.isNumber = isNumber;
/**
 * checks if the object provided is a string.
 * @param o
 */
function isString(o) {
    return isTrue(typeof o === "string");
}
exports.isString = isString;
/**
 * checks if the object provided is an array.
 * @param o
 */
function isArray(o) {
    return isTrue(Array.isArray(o));
}
exports.isArray = isArray;
/**
 * checks if the object provided is an object.
 * @param o
 */
function isObject(o) {
    return isTrue(typeof o === "object");
}
exports.isObject = isObject;
/**
 * checks if the object provided is null.
 * @param o
 */
function isNull(o) {
    return isTrue(o === null);
}
exports.isNull = isNull;
/**
 * checks if the object provided is defined, different of `undefined`.
 *
 * In Javascript land, `null` is considered a defined object.
 * @param o
 */
function isDefined(o) {
    return isTrue(typeof o !== "undefined");
}
exports.isDefined = isDefined;
